frontend/src/main/scala/bloop/io/DirChecksum.scala

```diff
@@ -0,0 +1,106 @@
+ */
+final class DirChecksum(dir: AbsolutePath,
+                        pattern: String,
+                        modifiedTimes: List[(AbsolutePath, FileTime)],
+                        contentsChecksum: Long) {
+
+  /**
+   * Inspects the directory for changes.
+   *
+   * @return true if the directory has changed, false otherwise.
+   */
+  def changed(): DirChecksum.DirStatus = {
+    val newModifiedTimes = DirChecksum.getFiles(dir, pattern)
+    if (newModifiedTimes == modifiedTimes) DirChecksum.DirUnchanged(None)
+    else {
+      val newChecksum = DirChecksum.filesChecksum(newModifiedTimes.map(_._1))
+      if (newChecksum != contentsChecksum) DirChecksum.DirChanged
+      else {
+        val checksum = new DirChecksum(dir, pattern, newModifiedTimes, newChecksum)
+        DirChecksum.DirUnchanged(Some(checksum))
+      }
+    }
+  }
+}
+
+object DirChecksum {
+
+  /** Indicates the status of a tracked directory */
+  sealed trait DirStatus
+
+  /**
+   * Indicates that the content of the directory hasn't changed.
+   *
+   * @param newChecksum If set, the new `DirChecksum` with the updated
+   *                    `lastModifiedTimes` (if the files have been touched, but
+   *                    not modified).
+   */
+  case class DirUnchanged(newChecksum: Option[DirChecksum]) extends DirStatus
+
+  /** Indicates that the content of the directory has changed. */
+  case object DirChanged extends DirStatus
+
+  /**
+   * Creates a new `DirChecksum`.
+   *
+   * @param dir     The directory to track.
+   * @param pattern The pattern that must be matched by the tracked files.
+   */
+  def apply(dir: AbsolutePath, pattern: String): DirChecksum = {
+    val files = getFiles(dir, pattern)
+    val checksum = filesChecksum(files.map(_._1))
+    new DirChecksum(dir, pattern, files, checksum)
+  }
+
+  /**
+   * Returns all the tracked files inside this directory, associated with their last
+   * modification time.
+   *
+   * @param dir     The directory to track.
+   * @param pattern The pattern to find the files to track.
+   * @return A map associating each tracked file with its last modification time.
+   */
+  private def getFiles(dir: AbsolutePath, pattern: String): List[(AbsolutePath, FileTime)] = {
+    Paths
+      .getAll(dir, pattern)
+      .map { path =>
+        path -> Files.getLastModifiedTime(path.underlying)
+      }
+      .toList
+  }
+
+  /**
+   * Computes the checksum of `files`.
+   *
+   * @param files The files whose checksum must be computed.
+   * @return The checksum.
+   */
+  private def filesChecksum(files: List[AbsolutePath]): Long = {
+    val streams = new Enumeration[InputStream] {
+      val streams = files.map(f => Files.newInputStream(f.underlying)).toIterator
+      override def hasMoreElements(): Boolean = streams.hasNext
+      override def nextElement(): InputStream = streams.next()
+    }
+
+    val checkedStream = new CheckedInputStream(new SequenceInputStream(streams), new Adler32)
```

Comment:
Can you elaborate on the contract of `CheckedInputStream` and `SequenceInputStream`? Do they immediately create a signature out of an input stream given a concrete algorithm? Is there any detail we should be particularly wary of?

Reply:
> Can you elaborate on the contract of CheckedInputStream and SequenceInputStream?

`CheckedInputStream` keeps an internal value that represents the checksum and that is updated as the wrapped stream is read through. `SequenceInputStream` reads several streams back to back. Closing the `CheckedInputStream` closes the underlying stream, and closing the `SequenceInputStream` closes all the underlying streams.

> Do they immediately create a signature out of an input stream given a concrete algorithm?

Exactly.

> Is there any detail we should be particularly wary of?

None that I know of.